signer/src/json.rs

```diff
@@ -0,0 +1,204 @@
+        match self {
+            Error::UnsupportedEncoding => write!(f, "Unsupported encoding."),
+            Error::Base64(e) => write!(f, "Base64 decoding error: {e}"),
+            Error::UnsupportedScryptParameters { n, p, r } => {
+                write!(f, "Unsupported Scrypt parameters: N: {n}, p: {p}, r: {r}")
+            }
+            Error::Secretbox(e) => write!(f, "Decryption error: {e}"),
+            Error::InvalidKeys => write!(f, "The decrypted keys are not valid."),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct EncryptedJsonDescriptor {
+    /// Descriptor for the content
+    content: Vec<String>,
+    /// The encoding (in current/latest versions this is always an array)
+    r#type: Vec<String>,
+    /// The version of encoding applied
+    version: String,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+struct Meta {
+    genesis_hash: String,
+    is_hardware: Option<bool>,
+    name: String,
+    tags: Option<Vec<String>>,
+    when_created: u64,
+}
+
+/// https://github.com/polkadot-js/common/blob/37fa211fdb141d4f6eb32e8f377a4651ed2d9068/packages/keyring/src/types.ts#L67
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+pub struct KeyringPairJson {
+    /// The encoded string
+    encoded: String,
+    /// The encoding used
+    encoding: EncryptedJsonDescriptor,
+    /// The ss58 encoded address or the hex-encoded version (the latter is for ETH-compat chains)
+    address: AccountId32,
+    /// The underlying metadata associated with the keypair
+    meta: Meta,
+}
+
+impl KeyringPairJson {
+    /// Decrypt JSON keypair.
+    pub fn decrypt(self, password: &str) -> Result<sr25519::Keypair, Error> {
+        // Check encoding.
+        // https://github.com/polkadot-js/common/blob/37fa211fdb141d4f6eb32e8f377a4651ed2d9068/packages/keyring/src/keyring.ts#L166
+        if self.encoding.version != "3"
+            || !self.encoding.content.contains(&"pkcs8".to_owned())
+            || !self.encoding.content.contains(&"sr25519".to_owned())
+            || !self.encoding.r#type.contains(&"scrypt".to_owned())
+            || !self
+                .encoding
+                .r#type
+                .contains(&"xsalsa20-poly1305".to_owned())
+        {
+            return Err(Error::UnsupportedEncoding);
+        }
+
+        // Decode from Base64.
+        let decoded = base64::engine::general_purpose::STANDARD.decode(self.encoded)?;
+        if decoded.len() < 68 {
+            return Err(Error::UnsupportedEncoding);
+        }
+
+        // Extract scrypt parameters.
+        // https://github.com/polkadot-js/common/blob/master/packages/util-crypto/src/scrypt/fromU8a.ts
+        let salt = &decoded[0..32];
+        let n = u32::from_le_bytes(decoded[32..36].try_into().unwrap());
```

Comment:
return err here instead unwrap please, perhaps add another error variant for these scrypt params.

Reply:
given that we check `decoded.len() < 68` above, I don't think it's necessary to also handle these errors as they should be infallible right? But I'd make them `.expect("length checked above")` perhaps :)

Reply:
I improved the error handling.